Add Cloudflare R2 as uploader module
Index: akkoma/lib/pleroma/uploaders/s3.ex
===================================================================
--- akkoma.orig/lib/pleroma/uploaders/s3.ex
+++ akkoma/lib/pleroma/uploaders/s3.ex
@@ -33,14 +33,12 @@ defmodule Pleroma.Uploaders.S3 do
         upload.tempfile
         |> ExAws.S3.Upload.stream_file()
         |> ExAws.S3.upload(bucket, s3_name, [
-          {:acl, :public_read},
           {:content_type, upload.content_type}
         ])
       else
         {:ok, file_data} = File.read(upload.tempfile)
 
         ExAws.S3.put_object(bucket, s3_name, file_data, [
-          {:acl, :public_read},
           {:content_type, upload.content_type}
         ])
       end
